File: python/Vector3.py
import math as m


class Vector3:
    def __init__(self, x=0, y=0, z=0):
        self.x = x
        self.y = y
        self.z = z

    def __str__(self):
        """
        Print overload.
        """
        return f"Vector3({self.x},{self.y},{self.z})"

    def __add__(self, vec):
        """
        Sum overload.
        """
        x = self.x + vec.x
        y = self.y + vec.y
        z = self.z + vec.z
        return Vector3(x, y, z)

    def __sub__(self, vec):
        """
        Substraction overload.
        """
        x = self.x - vec.x
        y = self.y - vec.y
        z = self.z - vec.z
        return Vector3(x, y, z)

    @classmethod
    def buildDir(self, pt1, pt2):
        return Vector3(pt2.x - pt1.x, pt2.y - pt1.y, pt2.z - pt1.z)

    def dot(self, vec):
        """
        Dot product.
        """
        return self.x * vec.x + self.y * vec.y + self.z * vec.z

    def cross(self, vec):
        """
        Cross product.
        """
        x = self.y * vec.z - self.z * vec.y
        y = self.z * vec.x - self.x * vec.z
        z = self.x * vec.y - self.y * vec.x
        return Vector3(x, y, z)

    def times(self, t):
        """
        Product with a scalar.
        """
        x = t * self.x
        y = t * self.y
        z = t * self.z
        return Vector3(x, y, z)

    def norm(self):
        return m.sqrt(self.x ** 2 + self.y ** 2 + self.z ** 2)

    def normalized(self):
        return self.times(1 / self.norm())

    def dist(self, pt):
        vec = Vector3(self.x - pt.x, self.y - pt.y, self.z - pt.z)
        return vec.norm()

File: rust/src/Vector3.rs
use std::ops::{Add, Sub, Mul};


#[derive(Clone, Copy)]
struct Vector3 {
    pub x : f64,
    pub y : f64,
    pub z : f64
}


impl Vector3 {
    pub fn print(self) {
        println!("Vector3({}, {}, {})", self.x, self.y, self.z);
    }

    pub fn buildDir(pt1 : Vector3, pt2 : Vector3) -> Vector3 {
        Vector3{
            x : pt2.x - pt1.x,
            y : pt2.y - pt1.y,
            z : pt2.z - pt1.z
        }
    }
    
    pub fn dot(self, v : Vector3) -> f64 {
        self.x * v.x + self.y * v.y + self.z * v.z
    }

    pub fn cross(self, v : Vector3) -> Vector3 {
        Vector3{
            x : self.y * v.z - self.z * v.y,
            y : self.z * v.x - self.x * v.z,
            z : self.x * v.y - self.y * v.x
        }
    }

    pub fn norm(self) -> f64 {
        (self.x.powi(2) + self.y.powi(2) + self.z.powi(3)).sqrt()
    }

    pub fn normalized(self) -> Vector3 {
        self * self.norm()
    }

    pub fn dist(self, pt : Vector3) -> f64 {
        Vector3::buildDir(self, pt).norm()
    }
}


// Vector sum
impl Add for Vector3 {
    type Output = Vector3;

    fn add(self, rhs : Self) -> Self::Output {
        Vector3{
            x : self.x + rhs.x,
            y : self.y + rhs.y,
            z : self.z + rhs.z
        }
    }
}


// Vector substraction
impl Sub for Vector3 {
    type Output = Vector3;

    fn sub(self, rhs : Self) -> Self::Output {
        Vector3{
            x : self.x - rhs.x,
            y : self.y - rhs.y,
            z : self.z - rhs.z
        }
    }
}


// Scalar product
impl Mul<f64> for Vector3 {
    type Output = Vector3;

    fn mul(self, rhs : f64) -> Self::Output {
        Vector3{
            x : self.x * rhs,
            y : self.y * rhs,
            z : self.z * rhs
        }
    }
}
